lib/active_stash/index_dsl.rb

```diff
@@ -0,0 +1,285 @@
+
+      unless [:has_one, :belongs_to].include?(reflection.macro)
+        raise ConfigError, "Only 1-to-1 associations (belongs_to and has_one) are currently supported"
+      end
+
+      dsl.instance_eval(&block)
+      association_indexes = dsl.finalize!.indexes
+      @indexes.concat(association_indexes.indexes.map do |idx|
+        idx.tap { |i| i.name = "#{path.join(".")}.#{i.name}" }
+      end)
+
+
+      inverse_name = reflection.inverse_of.name
+
+      @callback_registration_handlers.push(-> {
+        reflection.klass.after_save do |record|
+          record.send(inverse_name).try(:cs_put)
+        end
+
+        reflection.klass.after_destroy do |record|
+          # This will reindex every stash index associated with the model, not
+          # only the index that needs updating.  We can get smarter about this
+          # in the future.
+          record.send(inverse_name).reload
+          record.send(inverse_name).try(:cs_put)
+        end
+      })
+    end
+
+    private
+
+    def validate_no_fields_with_duplicate_index_definitions!
+      @indexes.group_by{|idx| idx.field }.each do |field, indexes|
+        indexes.group_by{|idx| idx.type }.each do |index_type, indexes|
+          if indexes.count > 1
+            raise ConfigError, "Multiple indexes of the same type on the same attribute: #{@model_class}##{field}, index type: #{index_type}"
+          end
+        end
+      end
+    end
+
+    def validate_no_association_duplicates!
+      # TODO
+    end
+
+    # RULES:
+    #
+    # 1. If no exact or range index already exists on the field, `unique` will
+    #    create a new exact index with a unique constraint.
+    #
+    # 2. Existing exact and range indexes for the field will be modified to
+    #    enforce a unique constraint, subject to the following rules:
+    #
+    # 2a. If the field is of type text or string, the unique constraint will
+    #     only be placed on the exact index.
+    #
+    # 2b. For every other field type, the unique constraint will be applied to
+    #     both range and exact indexes.
+    #
+    def process_unique_indexes!
+      lookup = ActiveStash::IndexLookup.new(@indexes.clone)
+
+      @unique_fields.each do |uf|
+        indexes_defined_on_field = lookup.on(uf.to_s).select{|idx| [:exact, :range].include?(idx.type) }
+        if indexes_defined_on_field.size == 0
+          # RULE 1.
+          @indexes.push(Index.exact(uf.to_s, unique: true))
+        else
+          # RULE 2.
+          field_type = @reflector.fields[uf.to_s]
+          if [:string, :text].include?(field_type)
+            # RULE 2a.
+            exact_index = indexes_defined_on_field.find{|idx| idx.type == :exact }
+            if exact_index
+              exact_index.make_unique
+            else
+              @indexes.push(Index.exact(field, unique: true))
+            end
+          else
+            # RULE 2b.
+            indexes_defined_on_field.each do |idx|
+              idx.make_unique
```

Comment:
Tiny thing. Calling this `make_unique!` might make it clearer that it is changing internal state.

Reply:
👍